govcd/tm_content_library_item.go

```diff
@@ -0,0 +1,377 @@
+func uploadContentLibraryItemFile(name string, cli *ContentLibraryItem, filesToUpload []*types.ContentLibraryItemFile, localFilePath string) error {
+	if cli == nil || len(filesToUpload) == 0 {
+		return fmt.Errorf("the Content Library Item or its files cannot be nil / empty")
+	}
+
+	// We just want to upload the selected file (named after the 'name' input parameter)
+	var fileToUpload *types.ContentLibraryItemFile
+	for _, f := range filesToUpload {
+		if f.Name == name {
+			fileToUpload = f
+			break
+		}
+	}
+	if fileToUpload == nil {
+		return fmt.Errorf("'%s' not found among the Content Library Item '%s' files", name, cli.ContentLibraryItem.Name)
+	}
+	filesAbsPaths, tmpDir, err := util.Unpack(localFilePath)
+	if err != nil {
+		return fmt.Errorf("%s. Unpacked files for checking are accessible in: %s", err, tmpDir)
+	}
+	defer func() {
+		err = os.RemoveAll(tmpDir)
+		if err != nil {
+			util.Logger.Printf("[DEBUG] could not clean up tmp directory %s", tmpDir)
+		}
+	}()
+
+	ud := uploadDetails{
+		uploadLink:               strings.ReplaceAll(fileToUpload.TransferUrl, "/transfer", "/tm/transfer"), // TODO: TM: Workaround, the link is missing /tm in path, so it gives 404 as-is
+		uploadedBytes:            0,
+		fileSizeToUpload:         fileToUpload.ExpectedSizeBytes,
+		uploadPieceSize:          1024,
+		uploadedBytesForCallback: 0,
+		allFilesSize:             fileToUpload.ExpectedSizeBytes,
+		callBack: func(bytesUpload, totalSize int64) {
+			util.Logger.Printf("[DEBUG] Uploaded Content Library Item file '%s': %d/%d", name, bytesUpload, totalSize)
+		},
+		uploadError: addrOf(fmt.Errorf("error uploading Content Library Item file '%s'", name)),
+	}
+
+	// When TM asks for a file called 'descriptor.ovf', it can be that inside the OVA
+	// it is not named like that. We search for an .ovf file in this case and we use it
+	foundName := name
+	if name == "descriptor.ovf" {
+		for _, f := range filesAbsPaths {
+			if filepath.Ext(f) == ".ovf" {
+				_, foundName = filepath.Split(f)
+				break
+			}
+		}
+	}
+
+	_, err = uploadFile(cli.client, findFilePath(filesAbsPaths, foundName), ud)
+	if err != nil {
+		return fmt.Errorf("could not upload the file: %s", err)
+	}
+	return nil
+}
+
+// getContentLibraryItemUploadTask searches for the task associated to the given Content Library Item upload and runs the input
+// function on it
+func getContentLibraryItemUploadTask(cli *ContentLibraryItem, operation func(task *Task) error) error {
+	taskRecords, err := cli.client.QueryTaskList(map[string]string{
+		"name":       "contentLibraryItemUpload",
+		"status":     "running,preRunning,queued,error",
+		"objectType": "contentLibraryItem",
+		"objectName": cli.ContentLibraryItem.Name,
+	})
+	if err != nil {
+		return err
+	}
+	var task *Task
+	for _, tr := range taskRecords {
+		if strings.Contains(tr.Object, cli.ContentLibraryItem.ID) {
```

Comment:
Not that much important, but possibly `equalIds()` can be useful :)

Reply:
Changed, thanks!